src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DoNotCallEnumerableCastThatWillFailAnalyzerTests.cs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Testing;
+using Xunit;
+using VerifyCS = Test.Utilities.CSharpCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Runtime.DoNotCallEnumerableCastThatWillFailAnalyzer,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+using VerifyVB = Test.Utilities.VisualBasicCodeFixVerifier<
+    Microsoft.NetCore.Analyzers.Runtime.DoNotCallEnumerableCastThatWillFailAnalyzer,
+    Microsoft.CodeAnalysis.Testing.EmptyCodeFixProvider>;
+
+namespace Microsoft.NetCore.Analyzers.Runtime.UnitTests
+{
+    public class DoNotCallEnumerableCastThatWillFailAnalyzerTests
+    {
+        [Fact]
+        public async Task DiagnosticCasesCSharp()
+        {
+            var ofTypeRule = DoNotCallEnumerableCastThatWillFailAnalyzer.OfTypeRule;
+
+            await VerifyCS.VerifyAnalyzerAsync(@"
+using System.Linq;
+
+class C
+{
+    public void M()
+    {
+        var a = (new int[0]).OfType<object>();
+        var b = (new int[0]).OfType<string>();
+        var c = (new object[0]).OfType<string>();
```

Comment:
Could we have tests for `.Cast<T>()`?

Reply:
Would also be good to have a test for `.Cast<T>` in [query expression syntax](https://docs.microsoft.com/en-US/dotnet/csharp/programming-guide/concepts/linq/converting-data-types)

```CS
_ = from string s in new object[0]
    select s;
```

Reply:
added some for both